core/domain/user_jobs_continuous_test.py

```diff
@@ -591,3 +597,458 @@ def test_basic_computation_works_if_collection_is_deleted(self):
+        }
+        return current_completions[exp_id]
+
+    @classmethod
+    def _mock_get_zero_impact_score(cls, exploration_id):
+        return 0
+
+    @classmethod
+    def _mock_get_below_zero_impact_score(cls, exploration_id):
+        return -1
+
+    def _run_manager(self):
+        """Runs the MapReduce job after running the continuous
+        statistics aggregator for explorations to get the correct num
+        completion events."""
+        with self.swap(stats_jobs_continuous.StatisticsAggregator,
+                    'get_statistics', self._mock_get_statistics):
+                job_id = ModifiedUserImpactAggregator.start_computation()
+                self.process_and_flush_pending_tasks()
+
+    def _run_exp_impact_calculation_and_assert_equals(
+            self, exploration_id, expected_impact):
+        with self.swap(stats_jobs_continuous.StatisticsAggregator,
+                    'get_statistics', self._mock_get_statistics):
+            self.assertEqual(expected_impact,
+                self.impact_mr_job_manager._get_exp_impact_score(
+                exploration_id))
+
+    def _sign_up_user(self, user_email, username):
+        # Sign up a user, have them create an exploration.
+        self.signup(user_email, username)
+        return self.get_user_id_from_email(user_email)
+
+    def _create_exploration(self, exp_id, user_id):
+        exploration = exp_domain.Exploration.create_default_exploration(
+            exp_id, 'A title', 'A category')
+        exp_services.save_new_exploration(user_id, exploration)
+        return exploration
+
+    def _rate_exploration(self, exp_id, num_ratings, rating):
+        """Create num_ratings ratings for exploration with exp_id,
+        of value rating.
+        """
+        # Each user id needs to be unique since each user can only give an
+        # exploration one rating.
+        user_ids = ['user{}'.format(i) for i in range(num_ratings)]
+        for user_id in user_ids:
+            rating_services.assign_rating_to_exploration(
+                user_id, exp_id, rating
+            )
+
+    def _complete_exploration(self, exploration, num_completions):
+        """Log a completion of exploration with id exp_id num_completions
+        times."""
+        exp_version = 1
+        state = exploration.init_state_name
+        session_ids = ['session{}'.format(i) for i in range(num_completions)]
+        for session_id in session_ids:
+            event_services.StartExplorationEventHandler.record(
+                exploration.id, exp_version, state, session_id, {},
+                feconf.PLAY_TYPE_NORMAL)
+            event_services.CompleteExplorationEventHandler.record(
+                exploration.id, exp_version, state, session_id, 27, {},
+                feconf.PLAY_TYPE_NORMAL)
+        # Set the number of completions, so mock can function
+        # correctly.
+        self.num_completions[exploration.id] = num_completions
```

Comment:
Should this be +=, in case a test wants to add two separate batches of completions to each exploration? If so, could do

    num_completions = collections.defaultdict(int)

and then you can do num_completions[exp_id] += ... without needing to check whether exp_id is a valid key.